corehq/apps/userreports/sql/adapter.py

```diff
@@ -48,20 +48,26 @@ def save(self, doc):
         indicator_rows = self.config.get_all_values(doc)
         if indicator_rows:
             table = self.get_table()
-            with self.engine.begin() as connection:
-                # delete all existing rows for this doc to ensure we aren't left with stale data
-                delete = table.delete(table.c.doc_id == doc['_id'])
-                connection.execute(delete)
-                for indicator_row in indicator_rows:
-                    all_values = {i.column.database_column_name: i.value for i in indicator_row}
-                    insert = table.insert().values(**all_values)
-                    try:
-                        connection.execute(insert)
-                    except IntegrityError:
-                        # Someone beat us to it. Concurrent inserts can happen
-                        # when a doc is processed by the celery rebuild task
-                        # at the same time as the pillow.
-                        pass
+            try:
+                with self.engine.begin() as connection:
+                    # delete all existing rows for this doc to ensure we aren't left with stale data
+                    delete = table.delete(table.c.doc_id == doc['_id'])
+                    connection.execute(delete)
+                    bad_saves = False
+                    for indicator_row in indicator_rows:
+                        all_values = {i.column.database_column_name: i.value for i in indicator_row}
+                        insert = table.insert().values(**all_values)
+                        try:
+                            connection.execute(insert)
+                        except (IntegrityError, InternalError):
+                            bad_saves = True
+                    if bad_saves:
+                        raise BadSaveError()
```

Comment:
is the purpose of this exception to rollback the transaction of the `with`-block? If that's the case, you could avoid the `BadSaveError` and some indentation levels with something like this:
```python
connection = engine.connect()
trans = connection.begin()
try:
    ...
except:
    trans.rollback()
    raise
if bad_saves:
    # discard bad saves silently
    # once we have a better strategy for showing these to users we can
    # revisit whether we should log something here.
    trans.rollback()
else:
    trans.commit()
```
It does seem troubling that you can call `indicator_thing.save(doc)` and it will return as if everything is fine when in fact the save failed.